Refactor resource listers

svenwltr opened the pull request:
Some heavy refactoring in preparation of #54. Also I want to get rid of the weird reflections.

Since this is a heavy change, here some notes:

This changes the interface for registering resources. Previously it was necessary to update the `GetListers` function in `resources/listers.go`. Now every resource registers itself inside of an `init` function which is defined in `resources/interface.go` (eg `register("IAMUser", ListIAMUsers)`). Since it has to provide a name for registering it isn't required anymore to get the resource type via reflections.

Also we are getting rid of the `FooNuke` types in `resources/types.go`, which just carried the session and don't really give a benefit. Now, the `ResourceLister` interface changed so it takes a session as a command. Therefore the intermediate types aren't required anymore.

svenwltr commented:
@rebuy-de/prp-aws-nuke Please review. Please do it carefully, since this is a big change.

bjoernhaeuser commented on cmd/queue.go:
```diff
@@ -15,32 +15,34 @@ const (
 
 type Item struct {
```
Should we start adding Godoc?

svenwltr replied:
I added a very bit of GoDoc

stephanlindauer commented on resources/listers.go:
```diff
@@ -1,112 +0,0 @@
-package resources
```
:heart:

bjoernhaeuser commented on resources/interface.go:
```diff
@@ -11,3 +15,13 @@ type Filter interface {
+var resourceListers = make(ResourceListers)
+
+func register(name string, lister ResourceLister) {
+	resourceListers[name] = lister
```
would prevent adding the same twice here

svenwltr replied:
done

bjoernhaeuser commented on resources/kms-aliases.go:
```diff
@@ -12,16 +13,22 @@ type KMSAlias struct {
-func (n *KMSNuke) ListAliases() ([]Resource, error) {
-	resp, err := n.Service.ListAliases(nil)
+func init() {
+	register("KMSAliase", ListKMSAliases)
```
i dont think that `Aliase` is incorrect :)

svenwltr replied:
fixed

bethge commented on resources/ec2-eip.go:
```diff
@@ -1,24 +1,33 @@
 
-func (n *EC2Nuke) ListAddresses() ([]Resource, error) {
+func init() {
+	register("EC2Addresse", ListEC2Addresses)
```
ping

bethge commented on resources/ecr-repository.go:
```diff
@@ -12,7 +13,13 @@ type ECRRepository struct {
 
-func (n *ECRNuke) ListRepos() ([]Resource, error) {
+func init() {
+	register("ECRRepo", ListECRRepos)
```
ping: `ECRRepository`

bethge commented on resources/elb-elb.go:
```diff
@@ -1,22 +1,31 @@
-func (n *ElbNuke) ListELBs() ([]Resource, error) {
-	resp, err := n.Service.DescribeLoadBalancers(nil)
+func init() {
+	register("ElbELB", ListElbELBs)
```
`ELB`?

bethge commented on cmd/scan.go:
```diff
@@ -8,14 +8,12 @@ import (
-
-		for _, lister := range listers {
-			r, err := safeLister(lister)
+		for category, lister := range resources.GetListers() {
```
This is by now the only place we use the variable name `category`. It is actually resource type name, right? Would love to get rid of `category` once and for all.

bethge commented on resources/elbv2-alb.go:
```diff
@@ -1,23 +1,32 @@
-func (n *Elbv2Nuke) ListELBs() ([]Resource, error) {
-	resp, err := n.Service.DescribeLoadBalancers(nil)
+func init() {
+	register("Elbv2ELB", ListElbv2ELBs)
```
`ELBv2`?

bethge commented on resources/elbv2-targetgroup.go:
```diff
@@ -1,23 +1,32 @@
-func (n *Elbv2Nuke) ListTargetGroups() ([]Resource, error) {
-	resp, err := n.Service.DescribeTargetGroups(nil)
+func init() {
+	register("Elbv2TargetGroup", ListElbv2TargetGroups)
```
`ELBv2TargetGroup`?

bethge commented on resources/elbv2-targetgroup.go:
```diff
@@ -1,23 +1,32 @@
+	register("Elbv2TargetGroup", ListElbv2TargetGroups)
+}
+
+func ListElbv2TargetGroups(sess *session.Session) ([]Resource, error) {
```
`ListELBv2TargetGroups`

bethge commented on resources/iam-policies.go:
```diff
@@ -10,8 +11,14 @@ type IAMPolicy struct {
-func (n *IAMNuke) ListPolicies() ([]Resource, error) {
-	resp, err := n.Service.ListPolicies(&iam.ListPoliciesInput{
+func init() {
+	register("IAMPolicie", ListIAMPolicies)
```
`IAMPolicy`

bethge commented on resources/kms-aliases.go:
```diff
@@ -12,16 +13,22 @@ type KMSAlias struct {
-func (n *KMSNuke) ListAliases() ([]Resource, error) {
-	resp, err := n.Service.ListAliases(nil)
+func init() {
+	register("KMSAliases", ListKMSAliases)
```
`KMSAlias`

bethge commented on resources/rds-subnets.go:
```diff
@@ -10,16 +11,22 @@ type RDSDBSubnetGroup struct {
 
-func (n *RDSNuke) ListSubnetGroups() ([]Resource, error) {
+func init() {
+	register("RDSSubnetGroup", ListRDSSubnetGroups)
```
`RDSDBSubnetGroup`

bethge commented on resources/route53-resource-records.go:
```diff
@@ -14,17 +15,23 @@ type Route53ResourceRecordSet struct {
 
-func (n *Route53Nuke) ListResourceRecords() ([]Resource, error) {
+func init() {
+	register("Route53ResourceRecord", ListRoute53ResourceRecords)
```
`Route53ResourceRecordSet`

svenwltr commented:
I added checked the resource names against version `1.3.0`. For that I added a new command `aws-nuke resource-types`. Since that command doesn't exist on `1.3.0`, I did some magic: `ack -oh '^type (\S+) struct' | grep -v Nuke | vi -`.

bethge commented on cmd/root.go:
```diff
@@ -70,6 +74,28 @@ func NewRootCommand() *cobra.Command {
+func NewResourceTypesCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "resource-types",
+		Short: "lists all availible resource types",
```
available

bethge commented on cmd/root.go:
```diff
@@ -70,6 +74,28 @@ func NewRootCommand() *cobra.Command {
+				types = append(types, resourceType)
+			}
+
+			sort.Sort(sort.StringSlice(types))
```
Would this work: https://golang.org/pkg/sort/#Strings ?
